src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -106,6 +136,150 @@ private void decideAndUseAPI(String remoteName, Item context, String credentials
         }
     }
 
+    /* Git repository URLs frequently end with the ".git" suffix.
+     * However, many repositories (especially https) do not require the ".git" suffix.
+     *
+     * Add remoteURL with the ".git" suffix and without the ".git" suffix to the
+     * list of alternatives.
+     */
+    private void addSuffixVariants(@NonNull String remoteURL, @NonNull Set<String> alternatives) {
+        alternatives.add(remoteURL);
+        String suffix = ".git";
+        if (remoteURL.endsWith(suffix)) {
+            alternatives.add(remoteURL.substring(0, remoteURL.length() - suffix.length()));
+        } else {
+            alternatives.add(remoteURL + suffix);
+        }
+    }
+
+    /* Git repository URLs frequently end with the ".git" suffix.
+     * However, many repositories (especially https) do not require the ".git" suffix.
+     *
+     * Add remoteURL with the ".git" suffix if not present
+     */
+    private String addSuffixVariant(@NonNull String canonicalURL) {
+        String suffix = ".git";
+        if (!canonicalURL.endsWith(suffix)) {
+            canonicalURL = canonicalURL + suffix;
+        }
+        return canonicalURL;
+    }
+
+    /* Protocol patterns to extract hostname and path from typical repository URLs */
+    private static Pattern gitProtocolPattern = Pattern.compile("^git://([^/]+)/(.+?)/*$");
+    private static Pattern httpProtocolPattern = Pattern.compile("^https?://([^/]+)/(.+?)/*$");
+    private static Pattern sshAltProtocolPattern = Pattern.compile("^[\\w]+@(.+):(.+?)/*$");
+    private static Pattern sshProtocolPattern = Pattern.compile("^ssh://[\\w]+@([^/]+)/(.+?)/*$");
+
+    /* Return a list of alternate remote URL's based on permutations of remoteURL.
+     * Varies the protocol (https, git, ssh) and the suffix of the repository URL.
+     * Package protected for testing
+     */
+    /* package */ @NonNull String convertToCanonicalURL(String remoteURL) {
+        if (remoteURL == null || remoteURL.isEmpty()) {
+            LOGGER.log(Level.FINE, "Null or empty remote URL not cached");
+            return ""; // return an empty string
+        }
+
+        Pattern [] protocolPatterns = {
+                gitProtocolPattern,
+                sshAltProtocolPattern,
+                sshProtocolPattern,
```

Comment:
I'd order these based on my assumption of most frequent use first, least frequent last.  The `sshAltProtocolPattern` is the pattern that GitHub and Bitbucket and Azure DevOps offer when they suggest an ssh URL.  Let's check it first.
```suggestion
                sshAltProtocolPattern,
                sshProtocolPattern,
                gitProtocolPattern,
```

Reply:
Ordering patterns on the basis of their frequency of usage seems like a nice optimisation, thanks for pointing this out.